---
output:
  word_document: default
  html_document: default
---
---
  title: "SQING Tables and Figures"
  output: word_document
  
  ---
  
  ```{r include = FALSE}
    #modified by DVM Bishop 4/2/17, 30/4/17
    
 
# 30th April 2017: moved this file to SQING_repo
  #-------------------------------------------------------------------------------
  # Specify packages
  #-------------------------------------------------------------------------------
  
  # Some of these packages may be unnecessary: I tried various options to allow word formatting
  #devtools::install_github("crsh/papaja")
  library(varhandle)
  library(papaja)
  library(XLConnect)
  library(Hmisc)
  library(knitr)
  library(pander)
   library(dplyr)
  library(gridExtra)
  library(grid)
  require(pwr)
  require(boot)
  require(psych) #for bootstrapped CI with cor.ci
  
  #see https://www.r-bloggers.com/r-markdown-how-to-insert-page-breaks-in-a-ms-word-document/
  # https://www.r-bloggers.com/tables-from-r-into-word/
  # http://stackoverflow.com/questions/37671868/knitr-rmarkdown-docx-tables/37890832#37890832
  # https://ourcodingclub.github.io/2016/11/24/rmarkdown-1.html -re miktex/mactex
  #-------------------------------------------------------------------------------
  
 #setwd("~/SQING_repo")
  myfilename<-"SQING_with_templates.xlsx" 
  
  
  SQING.dat=loadWorkbook(myfilename)
 #-------------------------------------------------------------------------------
 
# The organisation of information in the xls file is not optimal so we are going to 
# create a new data frame with column headings that make sense and then copy over
# the information we need
  
  #create a data.frame with 70 columns by 30 rows
  # includes some spare columns in case we need new variables
  
  ncols<-70
  nrows<-30
  allstudy<-data.frame(matrix(rep(NA,nrows*ncols),nrow=nrows)) #make data frame to hold all studies
  colnames(allstudy)=c('Study', 'Authors', 'Title', 'DOI', 'Sample', 'Sample_N', 'Subgroups', 
                       'Genes', 'Polymorphisms', 'Phenotypes', 'Analysis', 'SelResult', 'Conclusion', 
                       'SelRes_Variants_N', 'SelRes_ES', 'ES_source', 'Detect_ES_80', 'Power_ES_10', 
                       'Replication', 'Multcorr', 'Prevresults', 'Comment', 'Authorquery', 'Addinfo', 
                       'Geneticmodel', 'Haplotypes', 'N_Gmodels', 'CorrForN_Gmodels', 
                       'N_Polym', 'Correl_Polym', 'CorrForN_Polym','Genetic_analysis_notes', 'N_Pheno', 
                       'Correl_Pheno', 'CorrForN_Pheno', 'Neuro_Text', 'Neuro_CorrForN_approach','Imaging_Correction', 
                       'N_Neuropheno', 'CorrForN_Neuro', 'Mention_Multtest', 'Mention_ES', 
                       'Repcode', 'SelRes_N', 'Quasi_ES', 'Au_contacted', 'Au_email', 'Au_response', 
                       'Au_comments', 'Neuro_Method', 'Neuro_correct', 'SampleN_text', 'N_Subgroups', 'Journal', 'Combinations','X11','X12','X13','X14','X15',
                       'X16','X17','X19','X20')

# Note that some columns are not now used because we did not have time to collect sufficient information. 
# These include:
# Prevresults, Authorquery, Addinfo (this is blank)
# ------------------------------------------------------------------------------------------
#-------------------------------------------------------------------------------
   # Add data from each study, reading from the specific sheet for that study
#-------------------------------------------------------------------------------
 
  for (i in 1:30){
    mysheet<-paste('overview_',i,sep='')
  
  temp=data.frame(readWorksheet(SQING.dat,sheet=mysheet,startCol=2,endCol=2,header=F,useCachedValues=TRUE)) #most of what we need is in column B, so read that first
  maxrow<-nrow(temp)
  if(maxrow>60){maxrow=60} #some sheets have extra information:commentary from authors etc
  allstudy[i,1:maxrow]=t(temp)
  #move cols for text account of sample, and subgroups to cols 51 and 52
  #(historical reasons for this: these cols will be used for Ns for discovery/replication)
  allstudy[i,52]<-allstudy[i,6]
  allstudy[i,53]<-allstudy[i,7]

  #Add total N used to get selected result - in cells H4:H6
  temp2=data.frame(readWorksheet(SQING.dat,sheet=mysheet,startCol=8,endCol=8,startRow=4,endRow=6,header=F))
  allstudy[i,44]<-sum(temp2)
  
  temp3=data.frame(readWorksheet(SQING.dat,sheet=mysheet,startCol=4,endCol=5,startRow=22,endRow=22,header=F)) #Ns for discovery and replication in cells D22:E22
allstudy[i,6]<-temp3[,1]
allstudy[i,7]<-temp3[,2]

#categorical code for Neuro method is in cell C36
 temp4=data.frame(readWorksheet(SQING.dat,sheet=mysheet,startCol=3,endCol=3,startRow=36,endRow=36,header=F))
 allstudy[i,50]<-temp4[,1]

  }
colnames(allstudy)[6:7]<-c('DiscoveryN','ReplicationN')

# -----------------------------------------------------------------------------------------
# There was unexplained problem reading B15 (effect size) from some sheets 
# Affected sheets were 12, 17, 18, 20, 22, 30
# This has now been fixed by using the useCachedValues=TRUE setting with readWorksheet

# -----------------------------------------------------------------------------------------# Use depersonalised 3-category code for author response (from col D in 'authorcontact' worksheet)

temp5=data.frame(readWorksheet(SQING.dat,sheet='authorcontact',startCol=4,endCol=4,startRow=2,endRow=31,header=F))
 allstudy[,48]<-temp5[,1]
 table(allstudy$Au_response) #2 indicates replied; 1 indicates acknowledged,0 no response
 # These numbers to be entered into manuscript
 # ----------------------------------------------------------------------------------------

# Add journal names: these are in another data sheet

allstudy$Journal <-readWorksheet(SQING.dat,sheet='studies_with_replicates',startCol=3,endCol=3,header=T)

allstudy$Journal<-as.factor(unlist(allstudy$Journal))
# -----------------------------------------------------------------------------------------

# Add replication code

temp <- readWorksheet(SQING.dat,sheet='studies_with_replicates',startCol=9,endCol=9,header=T)
temp=temp[1:30,]
allstudy$Repcode<-as.factor(temp)
levels(allstudy$Repcode)<-c('No','Yes','Other','Prior')

# -----------------------------------------------------------------------------------------
# Just for tidiness, ensure all text columns have initial capitalisation
textcols=c(5,11,19,20,22,28,32,36,42)
for (k in 1:length(textcols)){
allstudy[,textcols[k]]<-capitalize(allstudy[,textcols[k]])
}
mytemp=which(allstudy$Genetic_analysis_notes==0) #tidying up text; replace 0 with - for this column
allstudy$Genetic_analysis_notes[mytemp]<-'-'
# -----------------------------------------------------------------------------------------

# Convert relevant columns into factors to allow tables

allstudy$Quasi_ES<-as.factor(allstudy$Quasi_ES)
#we'll add more relevant factor names once we've used this for plot

allstudy$Imaging_Correction<-as.factor(allstudy$Imaging_Correction)
levels(allstudy$Imaging_Correction)<-c('Whole area a priori ROI','Peak within a priori ROI', 'ROI from data','n/a')

allstudy$Correl_Polym=as.factor(allstudy$Correl_Polym) #whether polymorphisms correlated
allstudy$Correl_Pheno<-as.factor(allstudy$Correl_Pheno) #whether polymorphisms correlated
levels(allstudy$Correl_Polym)<-levels(allstudy$Correl_Pheno)<-c('NA','Yes','Unclear','No')

allstudy$CorrForN_Polym=as.factor(allstudy$CorrForN_Polym)
levels(allstudy$CorrForN_Polym)<-c('No','Bonferroni','Data Reduction','Permutation','NA')

allstudy$CorrForN_Pheno<-as.factor(allstudy$CorrForN_Pheno)
levels(allstudy$CorrForN_Pheno)<-c('None','Bonferroni', 'Permutation','Not needed')

allstudy$Au_response<-as.factor(allstudy$Au_response)
levels(allstudy$Au_response)<-c('None','Acknowledged','Responded')
# -----------------------------------------------------------------------------------------
# Recompute the power values corresponding to the SelRes_N
# -----------------------------------------------------------------------------------------
# (These have also been updated now in xls so should be the same )
for (i in 1:30){
  thisnum=as.integer(allstudy$SelRes_N[i])
  thispow=pwr.r.test(n=thisnum,sig.level =.05,power=.8,alternative="two.sided")
  result2= pwr.r.test(n =thisnum , sig.level =.05,r=.1,alternative="two.sided")
  thisr=thispow$r
  thisp=result2$power
  allstudy$Power_ES_10[i]=round(thisp,3)
  allstudy$Detect_ES_80[i]=round(thisr,3)
}


# -----------------------------------------------------------------------------------------
#Create a table with the Ns for multiple testing (Table 1 in paper)
# -----------------------------------------------------------------------------------------
myvars=c('Study','N_Subgroups','N_Polym','N_Gmodels','N_Pheno','N_Neuropheno','X16','Multcorr','Quasi_ES')
myNtable=allstudy[myvars]
mycombs<-as.integer(allstudy$N_Gmodels)*as.integer(allstudy$N_Polym)
temp<-as.integer(allstudy$N_Pheno)
temp[temp==0]<-1
mycombs<-mycombs*temp
temp<-as.integer(allstudy$N_Neuropheno)
temp[temp==0]<-1
mycombs<-mycombs*temp
temp<-as.integer(allstudy$N_Subgroups)
temp[temp==0]<-1
mycombs<-mycombs*temp

colnames(myNtable)[7]<-'Combinations'
allstudy$Combinations<-mycombs
myNtable$Combinations<-mycombs
#Now mark those where models, pheno or genos are correlated 

temploc<-which(myNtable$N_Gmodels>1)
temp<-myNtable$N_Gmodels[temploc]
temp=paste(temp,"-",sep='')
myNtable$N_Gmodels[temploc] <- temp  

temploc<-which(allstudy$Correl_Polym=='Yes')
temp<-myNtable$N_Polym[temploc]
temp=paste(temp,"-",sep='')
myNtable$N_Polym[temploc] <- temp  

#add a minus to those where Phenotypes correlated
temploc<-which(allstudy$Correl_Pheno=='Yes')
temp<-myNtable$N_Pheno[temploc]
temp=paste(temp,"-",sep='')
myNtable$N_Pheno[temploc] <- temp

#add a tilde to those where Phenotypes probably correlated
temploc<-which(allstudy$Correl_Pheno=='Unclear')
temp<-myNtable$N_Pheno[temploc]
temp=paste(temp,"~",sep='')
myNtable$N_Pheno[temploc] <- temp 
lastcol=length(colnames(myNtable))
myNtable<-myNtable[,-lastcol] #remove Quasi_ES column
# These codes allocated on basis of data in myNTable
myNtable$Corrected<-c(1,1,0,0,1,0,1,0,1,1,0,0,0,1,2,2,2,2,0,0,2,1,0,0,2,0,0,1,2,2)
myNtable$Corrected<-as.factor(myNtable$Corrected)
levels(myNtable$Corrected)<-c('No','Partial','Yes')

colnames(myNtable)<-c('Journal','Subgroups','Models','Polymorphisms','Phenotypes','Imaging regions','All Combinations','Correction method','Full correction')
#write.table(myNtable, "mynewtable.txt", sep="\t") #for copy on file

```


#Table 1
Multiple comparisons

```{r, echo=FALSE}


kable(myNtable,caption='Corrections for multiple comparisons in relation to N subgroups, genetic models, polymorphisms, and imaging regions. All combinations is the product of all of these. – denotes correlated variables; ~ denotes probably correlated
')
```
***
#Figure 2
##Obtained effect size in relation to sample size (on log scale)
##Symbols denote replication: triangle = Prior, square = Yes, circle = No, diamond = Other.

```{r, echo=FALSE} 
#NB black are 'quasi' effect sizes: need a legend to show this
# Also would be good to show correlation coeff for log sample size vs ES
temp=rep(0,30) #create vector to denote quasi effect sizes
myindex<-c(which(allstudy$Quasi_ES==0),which(allstudy$Quasi_ES=='n/a'))
mysubset<-allstudy[myindex,]
temp[myindex]<-1
plot(allstudy$SelRes_N,allstudy$SelRes_ES,log='x',
     xlab="Sample size ", ylab="Effect size (r)",
     pch=1+(15*temp)) #pch is symbol type: open or filled circle for 1 or 16
fullcorrel<-cor(log(as.numeric(allstudy$SelRes_N)),as.numeric(allstudy$SelRes_ES))
xy<-cbind(log(as.numeric(allstudy$SelRes_N)),as.numeric(allstudy$SelRes_ES))
fullci<-cor.ci(xy,niter=1000,plot=FALSE,p=.025)
subsetcorrel<-cor(log(as.numeric(mysubset$SelRes_N)),as.numeric(mysubset$SelRes_ES))
xy1<-cbind(log(as.numeric(mysubset$SelRes_N)),as.numeric(mysubset$SelRes_ES))
subsetci<-cor.ci(xy1,niter=1000,plot=FALSE,p=.025)

levels(allstudy$Quasi_ES)=c('No','Yes','n/a') #for Appendix
```

***
#Table 2 


```{r, echo=FALSE}

  temp=allstudy$N_Polym
temp2=which(temp>1)
temp[temp2]='Uncorrelated'
temp3=which(allstudy$Correl_Polym=="Yes")
temp[temp3]='Correlated'
tab1<-table(allstudy$CorrForN_Polym,temp)
nstudy<-sum(tab1[1:4,2:3])
kable(tab1[1:4,2:3],caption=paste('Correction for multiple testing in relation to genetic variants considered: ',nstudy,' studies with 2 or more correlated or uncorrelated polymorphisms',sep=''))
```

***

#Table 3


```{r, echo=FALSE}


tab3<-table(allstudy$CorrForN_Pheno,allstudy$Correl_Pheno)
kable(tab3,caption='Correction for multiple testing in relation to behavioural phenotypes')
```

***

#Table 4


```{r, echo=FALSE}
tab4<-table(allstudy$Journal,allstudy$Repcode)
kable(tab4,caption='Number of studies including replication sample, by journal')
```


***

#Appendix

```{r, echo=FALSE,results = 'asis'}
mytempx<-matrix(rep(NA,2*71),ncol=2)
mytempx[1:70,1]<-colnames(allstudy)# for checking col number of relevant columns
neworder<-c(3,4,54,13,5,6,7,8,9,29,30,10,33,34,39,11,25,27,50,12,16,44,22,15,45,38,41,28,31,35,37,17,18,48) #Authors omitted
mytempl<-matrix(rep(NA,2*34),ncol=2)

mytempl[1:34,1]<-c('Title', 'DOI', 'Journal','Conclusion','Sample', 'Discovery Sample N',
                   'Replication Sample N',
                   'Genes', 'Polymorphisms', 'N Polymorphisms','Correlated Polymorphisms',
                   'Phenotypes', 'N Behav. Phenotypes','Correlated Behav. Phenotypes',
                   'N Neuro Phenotypes','Analysis Method', 'Genetic models',
                   'N genetic models', 'Neuro methods','Selected Result',
                   'Selected Result Source','Ns for Selected Result',
                   'Comment on Selected Result','Selected Result Effect Size',
                   'Quasi Effect Size?','Imaging ROI approach',
                   'Mention of multiple comparisons','Correction for N genetic models',
                   'Correction for N polymorphisms','Correction for N behavioural phenotypes',
                   'Correction for N neuro phenotypes','Effect Size (r) Detectable with 80% Power',
                   'Power to Detect Effect Size (r) of .1','Author response to contact')
for (i in 1:30){
  thisstudy=paste('Study',i)
  #print(thisstudy)
  colnames(mytempl)=c('Information',thisstudy)
  mytempl[,2]<-t(allstudy[i,neworder])
print(kable(mytempl,caption='')) #Have to explicitly print if in a loop
 
 # print(grid.table(mytempl))
  #need to look at formatting with ftp://cran.r-project.org/pub/R/web/packages/gridExtra/vignettes/tableGrob.html

}

